File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        deque<int> dq;
        vector<int> result;

        for(int i=0; i<k; i++){
            while(!dq.empty() && dq.back() < nums[i]){
                dq.pop_back();
            }
            
            dq.push_back(nums[i]);
        }
        result.push_back(dq.front());
        for(int i=k; i<n; i++){
            
            if(dq.front() == nums[i-k]){
                dq.pop_front();
            }
            
            while(!dq.empty() && dq.back() < nums[i]){
                dq.pop_back();
            }
            
            dq.push_back(nums[i]);
            
            result.push_back(dq.front());
        }
        
        return result;
    }
};

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n = nums.length;
        int result[] = new int[n-k+1];
        Deque<Integer> dq = new ArrayDeque<>();
        int j = 0;
        for(int i=0; i<k; i++){
            while(!dq.isEmpty() && dq.peekLast() < nums[i]){
                dq.pollLast();
            }
            
            dq.addLast(nums[i]);
        }
        result[j++] = dq.peek();
        for(int i=k; i<n; i++){
            if(dq.peek() == nums[i-k]){
                dq.poll();
            }
            
            while(!dq.isEmpty() && dq.peekLast() < nums[i]){
                dq.pollLast();
            }
            
            dq.addLast(nums[i]);
            
            result[j++] = dq.peek();
        }
        return result;
    }
}